### ISSUE TITLE
LXC Assigning same cores in single container multiple times.

### ISSUE THREAD

**Issue by hnagri** [lxc/lxd#3109]
* Distributor ID:	Ubuntu
* Description:	Ubuntu 16.04 LTS
* Release:	16.04
* Codename:	xenial
 * The output of "lxc info" or if that fails:
```
 # lxc info
config:
  core.https_address: 0.0.0.0:8443
  core.trust_password: true
api_extensions:
- storage_zfs_remove_snapshots
- container_host_shutdown_timeout
- container_syscall_filtering
- auth_pki
- container_last_used_at
- etag
- patch
- usb_devices
- https_allowed_credentials
- image_compression_algorithm
- directory_manipulation
- container_cpu_time
- storage_zfs_use_refquota
- storage_lvm_mount_options
- network
- profile_usedby
- container_push
- container_exec_recording
- certificate_update
- container_exec_signal_handling
- gpu_devices
- container_image_properties
- migration_progress
- id_map
- network_firewall_filtering
- network_routes
- storage
- file_delete
- file_append
- network_dhcp_expiry
api_status: stable
api_version: "1.0"
auth: trusted
public: false
environment:
  addresses:
  - 10.1.248.3:8443
  - 10.1.249.3:8443
  architectures:
  - x86_64
  - i686
  certificate: |
    -----BEGIN CERTIFICATE-----
    MIIGATCCA+mgAwIBAgIQSpECGa64gzXQLPeODs+o+DANBgkqhkiG9w0BAQsFADBE
    MRwwGgYDVQQKExNsaW51eGNvbnRhaW5lcnMub3JnMSQwIgYDVQQDDBtyb290QGJi
    LWJsci1wcm9kLWNvbXB1dGUtMDMwHhcNMTcwMTEzMDkyNTI4WhcNMjcwMTExMDky
    NTI4WjBEMRwwGgYDVQQKExNsaW51eGNvbnRhaW5lcnMub3JnMSQwIgYDVQQDDBty
    b290QGJiLWJsci1wcm9kLWNvbXB1dGUtMDMwggIiMA0GCSqGSIb3DQEBAQUAA4IC
    DwAwggIKAoICAQDaSNSbODu2Bl1zz/hpjFbhvJXQM7hDPF9MO194Ybm/RiPns4Zb
    VzVhfziUuDUmJxJxrTSGO8i2f55RyXg280PX3jDka4cNCMI4vTTjGUWvwoVDpjeK
    UUU0DAe0kTCPgp1lKDmq3u/nmboGGFdRKkqMTxmCsNaxkhxryoYWoCCRs1k/+dLv
    cjfzHR+rsaXax0z1HlgXOCamjsUkId/XOTWi2o0HeMWbHRRV+BmxGfUBvfSS2LGp
    jd8T5aTReujNOxYwi3Qo3H0mqvcerIaoSg4JMXlvzvb2uGGjNrl22DYoXgXCx7RU
    MhU8Jg422Vwq0iUV+vzGjJDKyWszAANtvsep6YU2EPxPnM0idl/37zPtI9h8EiEa
    vfp76RDjjJN0Nf6Z1hmJj13gIa3+6jowEoI6GwT1Lpa5JmNbTgRdxWEiwsJm4Eo3
    ynsFDmZLa6kZRWsoveE3vw1BWbKwk0iqQunP3u1jTDohYNhKtLPcwfxqZwrrTWiE
    5liSOa9bW93Yrwz/9Cw/dT/0XN7XQV1lcS8PyQRmJqJlvKFu191RgI6eMFe1wBcd
    at00eUvV4uvgBbGCwkT/EF4KuwSzE94FxN5WrxeiI1j3OSS6/LEWPSEOvvuDDtYU
    ZnfH/9jm/TLQUg3GYSqMCyQs5CvrfDfThv4wWVAhfUbMpELkn502ibYfNwIDAQAB
    o4HuMIHrMA4GA1UdDwEB/wQEAwIFoDATBgNVHSUEDDAKBggrBgEFBQcDATAMBgNV
    HRMBAf8EAjAAMIG1BgNVHREEga0wgaqCFmJiLWJsci1wcm9kLWNvbXB1dGUtMDOC
    DTEwLjEuMjQ4LjMvMjSCG2ZlODA6OjI2NmU6OTZmZjpmZTBlOmIzMC82NIIcZmU4
    MDo6YTIzNjo5ZmZmOmZlYjY6OGViNC82NIIcZmU4MDo6YTIzNjo5ZmZmOmZlYjY6
    OGViNi82NIIcZmU4MDo6Nzg5ODo0N2ZmOmZlNzA6YTVmNi82NIIKZmU4MDo6MS82
    NDANBgkqhkiG9w0BAQsFAAOCAgEAxh4JIMr4XhxXfGnuZVMs+shA12e1rgyM8jBm
    xRBXB4YmHeapQ/TTFCbEL4xj4501TbpnDqceSkAdliysblZb3Gkjo1vHInWNaMw5
    OMcWzFj+fK05ZN6WrgrpjOBsfNsUjzGQM7O5xCoB7zV7ajus1YDmd+rwrbq+KXom
    aso2NMaBB7oDYO/yqVXD4mHRMrD7D3DZQiuoi/piKE2IFKBiL6BahCyWXVwISG06
    Icy421/szbwzaoiDQUYLXD5iNhiGs1SPihjq/bqy+2MXyZ/JixLbOo1BoQ/5KBrk
    4bq9DxyPgM+dDYwaWHIm0O8Noyt6GwNHX/T4DC2iHmPEq2lLdbBJ3x2evjFMWVLp
    8GTAgiOxIDUnOQiT5pWhrh+Qd8sJHXkuk1cDdQZ8P+lj8DwL01Uh4Lh85tFXg6zM
    2HK3I8BSqrChbP6YRwsxafr3rs0pm0UqlLMFTPDfxKsWLFI/C+AYIqrRf3iHM1wC
    dHTLdkF0MCDUBsQIU/bVQpy7tgNiqrGPhN/F2OeOjLj+PhdQhgr7HvDuQ/vIAf/j
    oSvHd3YUwJplEW4c67Q5Dkmnh1821ROnQqYHNvpdRo2iEWze+3iZlOIykypRBu5V
    t7iuylU=
    -----END CERTIFICATE-----
  certificate_fingerprint: c5d6625d2184aaedf6e6ed966e3465a9f2d8b2cd7f1733ac330a2b4b0702
  driver: lxc
  driver_version: 2.0.5
  kernel: Linux
  kernel_architecture: x86_64
  kernel_version: 4.4.0-59-generic
  server: lxd
  server_pid: 186992
  server_version: 2.9.3
  storage: zfs
  storage_version: 0.6.5.6-0ubuntu15

```

We create a lot of containers on a high-end host machine.
this is an example of one of such container.
following is container config:

```
# lxc config show B059APP8307220
architecture: x86_64
config:
  boot.autostart: "1"
  image.architecture: x86_64
  image.description: Ubuntu 16.04 LTS server (20160907.1)
  image.os: ubuntu
  image.release: xenial
  limits.cpu: "2"
  limits.memory: 8GB
  volatile.base_image: 86b8394ca69792791a095894863f468e2b999975fa92c2b0bb09a855b6189941
  volatile.eth0.hwaddr: 00:16:3e:bd:c4:96
  volatile.idmap.base: "0"
  volatile.idmap.next: '[{"Isuid":true,"Isgid":false,"Hostid":100000,"Nsid":0,"Maprange":65536},{"Isuid":false,"Isgid":true,"Hostid":100000,"Nsid":0,"Maprange":65536}]'
  volatile.last_state.idmap: '[{"Isuid":true,"Isgid":false,"Hostid":100000,"Nsid":0,"Maprange":65536},{"Isuid":false,"Isgid":true,"Hostid":100000,"Nsid":0,"Maprange":65536}]'
  volatile.last_state.power: RUNNING
devices:
  eth0:
    host_name: B059APP8307220
    name: eth0
    nictype: bridged
    parent: br0
    type: nic
  root:
    path: /
    pool: lxd
    size: 2GB
    type: disk
  shared_log_1:
    path: /DAS
    source: /gluster-compute/DAS
    type: disk
  shared_log_2:
    path: /LogBackups
    source: /LogBackups
    type: disk
ephemeral: false
profiles:
- default
```

You can see we have assigned 2 cpu cores to the container.
The Problem is when inside the container we do ;

```
root@B059APP8307220:~# cat /proc/cpuinfo
processor   : 0
vendor_id   : GenuineIntel
cpu family  : 6
model       : 79
model name  : Intel(R) Xeon(R) CPU E5-2683 v4 @ 2.10GHz
stepping    : 1
microcode   : 0xb00001e
cpu MHz     : 2793.902
cache size  : 40960 KB
physical id : 0
siblings    : 32
core id     : 1
cpu cores   : 16
apicid      : 3
initial apicid  : 3
fpu     : yes
fpu_exception   : yes
cpuid level : 20
wp      : yes
flags       : fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc aperfmperf eagerfpu pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch epb intel_pt tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm rdseed adx smap xsaveopt cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts
bugs        :
bogomips    : 4200.02
clflush size    : 64
cache_alignment : 64
address sizes   : 46 bits physical, 48 bits virtual
power management:

processor   : 1
vendor_id   : GenuineIntel
cpu family  : 6
model       : 79
model name  : Intel(R) Xeon(R) CPU E5-2683 v4 @ 2.10GHz
stepping    : 1
microcode   : 0xb00001e
cpu MHz     : 1266.890
cache size  : 40960 KB
physical id : 1
siblings    : 32
core id     : 1
cpu cores   : 16
apicid      : 35
initial apicid  : 35
fpu     : yes
fpu_exception   : yes
cpuid level : 20
wp      : yes
flags       : fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc aperfmperf eagerfpu pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch epb intel_pt tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm rdseed adx smap xsaveopt cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts
bugs        :
bogomips    : 4201.26
clflush size    : 64
cache_alignment : 64
address sizes   : 46 bits physical, 48 bits virtual
power management:

```

You can see that the **core id** of both the processor is same.

```
root@B059APP8307220:~# cat /proc/cpuinfo | grep 'core id'
core id		: 1
core id		: 1

```

Now this does not happen all the time but it does happen some times. 

My question is, is this normal? or is this a bug? cause when it happens the performance of the container downgrades too much.

**Comment by siteshbehera** [comment]
Following the writeup : 

Just give me X CPUs
In this mode, you let LXD pick a bunch of cores for you and then load-balance things as more containers and CPUs go online/offline.
The container only sees that number of CPU.

This works perfectly, but the core-id which are assigned is same. That means, LXC container shows up as 2 cores assigned, but actually its a single cpu assigned. 

This creates lot of performance issues when multi threaded applications looking to take advantage of multi cores, gets tied up to single core and races for cpu. 

Having said this,  please clarify if **limits.cpu:** be used for over spawning of containers ?  


Again we have observed that, even if we are not assigning more cpus to what a hypervisor has, same cpu id gets assigned to multiple containers. I think LXD should maintain a kind of resource management, where there should be flags to allow greedy mode or conservative mode. 

If conservative mode is flagged, LXD should ensure unique cpu ids are assigned to each container. It will be responsibility of the administrator to take care of not over committing cpus or LXD should throw an error in that case.

**Comment by brauner** [comment]
`LXD` has nothing to do with assigning the `core id`. The `core id` is used to identify which processors are on what core. If you only have one core but say two processors then `core id` will show the same number for both processors. So if you tell `LXD` give me two processors it will give you two and they can be on the same core or not. What you seem to want is a way to assign actual cores say give me the core with `core id 1`. That doesn't make much sense afaik.

**Comment by siteshbehera** [comment]
@brauner, Thx for clarification. I just now identified one thing, though core id is same, physical id is different. That means, same core id, but diff physical core. 

I will  be glad if you can clarify the behavior for  **limits.cpu** once more. 

Suppose, we have 16 physical cores per cpu and say there are 2 cpus. With hyperthreading enabled, and per container say I am putting **limits.cpu : 4** , does it mean , I can only spawn 15 containers on a hypervisor, considering 4 cores dedicated to hypervisor  itself.
If so, what will be LXD behavior when I spawn 2 more containers with 4 cores assigned ? 

In a nutshell how LXD behaves on overcommit using **limits.cpu**

**Comment by stgraber** [comment]
Right, I was about to point the original reporter to the fact that the physical id was different and so the container is pinned to the second core of the first CPU and second core of the second CPU.

**Comment by brauner** [comment]
I'm not sure what you mean how LXD would behave. Technically, you can spawn as many containers as you want. If you care about overcommitting then you can set `limits.cpu.priority` for each of your containers. This is documented in `doc/configuration.md`. You can set it to an integer between `0` and `10`. This will determine the CPU scheduling priority for this container relative to other containers sharing the same cpu{s}.

**Comment by stgraber** [comment]
So with a dual-socket, 16 cores, 2 threads mode, you'd have "64" CPUs as far as LXD (and the kernel scheduler) is concerned. So with 4 CPUs per container, you'll be able to spawn 16 containers before they start having to share CPUs.

LXD's scheduling (which drives the kernel's scheduler through cgroups) moves containers between CPUs so that each CPU has the least amount of containers. If you don't overcommit, then each CPU will have at most one container using it. If you do overcommit, then some CPUs will have 2 containers running on them (or more depending on how much you overcommit).

LXD will re-balance the containers every time a container is started or stopped, to try and keep a similar load on all CPUs. If your system supports CPU hotplug (unlikely since your system is already a populated 2 sockets), then hot plugging a new CPU will also cause LXD to re-balance.

As @brauner mentioned, there are additional options in doc/configuration.md to have LXD tweak the kernel's behavior when over-committing so you can give the kernel scheduler hints as to which container should get the CPU time first.

**Comment by siteshbehera** [comment]
@stgraber  @brauner : Thanks a lot for the clarification. Its much clear now. I will request you to please keep the bug open for some time, as I need to validate the same on my lab.  Appreciate your quick response.

**Comment by siteshbehera** [comment]
Hi, we verified the cpu core allotment. Following are the observations:

1) Rebalance works good when a container restarts.
2) If not over committed core-id and physical id is unique.

You can please go ahead and close the bug.
